clang/tools/driver/cc1_main.cpp

```diff
@@ -858,6 +971,255 @@ ObjectStoreCachingOutputs::replayCachedResult(llvm::cas::ObjectRef ResultID,
   return 0;
 }
 
+Expected<bool> RemoteCachingOutputs::tryReplayCachedResult(
+    const llvm::cas::CASID &ResultCacheKey) {
+  DiagnosticsEngine &Diags = Clang.getDiagnostics();
+
+  RemoteKVClient->getValueQueue().getValueAsync(ResultCacheKey.getHash());
+  Expected<remote_cache::KeyValueDBClient::GetValueAsyncQueue::Response>
+      Response = RemoteKVClient->getValueQueue().receiveNext();
+  if (!Response)
+    return Response.takeError();
+  if (!Response->Value) {
+    Diags.Report(diag::remark_compile_job_cache_miss)
+        << ResultCacheKey.toString();
+    return false;
+  }
+
+  Expected<bool> ReplayedResult =
+      replayCachedResult(ResultCacheKey, *Response->Value);
+  if (!ReplayedResult)
+    return ReplayedResult.takeError();
+
+  if (*ReplayedResult) {
+    Diags.Report(diag::remark_compile_job_cache_hit_resultless)
```

Comment:
What does it mean by having a resultless hit? And why is this a hit as you can't retrieve the output?

Reply:
I realize the name `remark_compile_job_cache_hit_resultless` is confusing..

The `remark_compile_job_cache_hit` diagnostic expects to print out the CASID association of "'\<key-casid>' => '\<result-casid>'", but in the case of the remote key-value request there is no "result-casid" to report, so I added `remark_compile_job_cache_hit_resultless` that only prints the "key-casid" (it is a remark diagnostic "without printing the result" / "resultless").

How about to rename to "remark_compile_job_cache_hit_only_key" ?

Reply:
I don't quite understand why there is no "result-casid"? Because the result id can only be decoded in replay function? Sounds like we should get that ID out of replay function.

Reply:
In the remote key-value API the value that you get is a dictionary (`message Value { map<string, bytes> entries = 1; }`), there is no single CASID to report.

Reply:
I know about that, I am just wondering if we should surface out the CASID for output in the dictionary.

Reply:
You mean dump out the whole dictionary? I don't think it's that useful, in practice I haven't encountered an actual need to use the printed result CASID. But we can re-evaluate if we find a compelling use, probably with a separate remark for enabling even more detailed/verbose diagnostic output.

Reply:
Do you mean just print the `bytes` for the entry with `"<output>"` key, ignoring the rest?

Reply:
Yup, maybe `replayCachedResult` can return `Expected<Optional<CASID>>`?

Reply:
Sounds good! See https://github.com/apple/llvm-project/pull/5355/commits/67709c67630a9cf21083e611b594554695c80cdb
I moved printing the remark in `replayCachedResult()`, which also helps for printing the remark before the cached stderr.